server/src/completion/YmlCompletionItemsProvider.ts

```diff
@@ -26,7 +26,23 @@ export class YmlCompletionItemsProvider {
     );
   }
 
-  public getOnlyCompilationItems() {
-      return this.completions.map((elem) => elem.completion);
+  public getAvailableCompletionItems(uri: string, offset: number) {
+    return this.completions
+      .map((elem) => {
+        if (elem.scopeEndOffset && elem.scopeStartOffset && elem.uri === uri) {
+          // We have the scope's positions set and are in the correct file.
+          if (offset < elem.scopeStartOffset || offset > elem.scopeEndOffset) {
+            // Outside of the scope.
+            return null;
+          } else {
+            return elem.completion;
+          }
+        } else {
+          // No information about the scope. The element is available everywhere.
+          return elem.completion;
```

Comment:
* It surprises me that you have two nested if-else with the very same “else” content. You could probably merge stuff. If it results in weird huge conditions, use functions (well, or just one or whatever) to make sure people understand what's happening.
* As a general rule, if your “else” is way shorter, invert your if-else to increase readability.